snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,648 @@
+        possible columns are missing from the select clause, then the ordering will
+        be not as expected.
+        """
+        for i, col in enumerate(self.__all_columns):
+            if column_name == col:
+                return LiteralExpr(None, i + 1)
+        raise ValueError(f"Unknown column: {column_name}")
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        array_joins_in_query = self.__get_array_joins_in_query(query)
+
+        tuple_alias = self.__get_unused_alias(query)
+
+        single_filtered, multiple_filtered = self.__get_filtered(query)
+
+        def replace_expression(expr: Expression) -> Expression:
+            match = self.array_join_pattern.match(expr)
+
+            # The arrayJoins we are looking for are not present, so skip this entirely
+            if match is None:
+                return expr
+
+            # All of the possible array joins are present
+            if array_joins_in_query == set(self.__all_columns):
+                tuple_index = self.__find_tuple_index(match.string("col"))
+
+                single_index_filtered = {
+                    self.__find_tuple_index(column_name): filtered
+                    for column_name, filtered in single_filtered.items()
+                }
+
+                multiple_indices_filtered = {
+                    tuple(
+                        self.__find_tuple_index(column) for column in column_names
+                    ): filtered
+                    for column_names, filtered in multiple_filtered.items()
+                }
+
+                if single_filtered or multiple_filtered:
+                    return filtered_mapping_tuples(
+                        expr.alias,
+                        tuple_alias,
+                        tuple_index,
+                        self.__all_columns,
+                        single_index_filtered,
+                        multiple_indices_filtered,
+                    )
+
+                return unfiltered_mapping_tuples(
+                    expr.alias, tuple_alias, tuple_index, self.__all_columns
+                )
+
+            # Only array join present is one of the key columns
+            elif len(array_joins_in_query) == 1 and any(
+                column in array_joins_in_query for column in self.__key_columns
+            ):
+                column_name = array_joins_in_query.pop()
+                if column_name in single_filtered:
+                    return filtered_mapping_keys(
+                        expr.alias, column_name, single_filtered[column_name]
+                    )
+
+            # No viable optimization
+            return expr
+
+        query.transform_expressions(replace_expression)
+
+        # If there is a bloom filter index present on the key columns, we want
+        # to take advantage of it by adding in the necessary `has` conditions.
+        if self.__use_bf_index:
```

Comment:
I think I would make the bloom filter check a separate query processor. It feels like it's just tacked on here and would simplify this processor a bit. Up to you but the arrayjoin optimization and the bloom filter optimization feel like separate concepts.

Reply:
Split out in de2187c